sync/fetch_queue.go

```diff
@@ -0,0 +1,315 @@
+package sync
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"sync"
+)
+
+type FetchPoetProofFunc func(poetProofRef []byte) error
+type GetValidAddressableTxFunc func(tx *types.SerializableSignedTransaction) (*types.AddressableSignedTransaction, error)
+type SValidateAtxFunc func(atx *types.ActivationTx) error
+type CheckLocalFunc func(ids []types.Hash32) (map[types.Hash32]Item, map[types.Hash32]Item, []types.Hash32)
+
+type Item interface {
+	Hash32() types.Hash32
+	ShortString() string
+}
+
+type fetchJob struct {
+	items []Item
+	ids   []types.Hash32
+}
+
+//todo make the queue generic
+type fetchQueue struct {
+	log.Log
+	BatchRequestFactory
+	*sync.Mutex
+	workerInfra WorkerInfra
+	pending     map[types.Hash32][]chan bool
+	handle      func(fj fetchJob)
+	checkLocal  CheckLocalFunc
+	queue       chan []types.Hash32 //types.TransactionId //todo make buffered
+}
+
+func (vq *fetchQueue) Close() {
+	vq.Info("done")
+	close(vq.queue)
+}
+
+//todo batches
+func (fq *fetchQueue) work() error {
+	output := fetchWithFactory(NewFetchWorker(fq.workerInfra, 1, fq.BatchRequestFactory, fq.queue))
+	for out := range output {
+		if out == nil {
+			fq.Info("close queue")
+			return nil
+		}
+
+		bjb, ok := out.(fetchJob)
+		if !ok {
+			fq.Error(fmt.Sprintf("Type conversion err %v", out))
+			continue
+		}
+
+		if len(bjb.ids) == 0 {
+			return fmt.Errorf("channel closed")
+		}
+
+		if len(bjb.items) == 0 {
+			fq.Warning("could not fetch any items ") //todo print ids
+			continue
+		}
+
+		fq.handle(out.(fetchJob))
+		fq.Debug("next batch")
+	}
+	return nil
+}
+
+func (fq *fetchQueue) addToPendingGetCh(ids []types.Hash32) chan bool {
+	return getDoneChan(fq.addToPending(ids))
+}
+
+func (fq *fetchQueue) addToPending(ids []types.Hash32) []chan bool {
+	fq.Lock()
+	deps := make([]chan bool, 0, len(ids))
+	var idsToAdd []types.Hash32
+	for _, id := range ids {
+		ch := make(chan bool, 1)
```

Comment:
Note that you are creating at least one channel for each id. If you have duplicated requests you will have multiple channels per id.
I think you should think of an alternative (more simple) mechanism or simplify it by having only one channel per item and record the number of "subscribers" per item so you can notify all of them (without requiring creating a channel to each one of them). Think of buffer limitations as it may suggest that this "simple" solution is not feasible.
Look into the option of having a wg per item collection (@see my comment on getDoneChan).

Reply:
i replayed above